datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,551 @@
+
+        let id_count = group_schema.fields().len();
+        let id_field = group_schema.fields().first().map(|f| f.data_type());
+        let ag: Box<dyn LimitedAggregator> =
+            match (id_count, id_field, val_field.data_type()) {
+                // TODO: solve combinatorial issue
+                (1, Some(DataType::Utf8), DataType::Int64) => {
+                    Box::new(StringInt64Aggregator::new(limit, descending)?)
+                }
+                _ => {
+                    panic!("Type not implemented: {:?} {:?}", id_field, val_field.data_type());
+                }
+            };
+
+        Ok(GroupedPriorityQueueAggregateStream {
+            schema: agg_schema,
+            input,
+            aggregate_arguments,
+            group_by: agg_group_by,
+            aggregator: ag,
+        })
+    }
+}
+
+impl RecordBatchStream for GroupedPriorityQueueAggregateStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+trait LimitedAggregator: Send {
+    fn intern(&mut self, ids: ArrayRef, vals: &[ArrayRef]) -> Result<()>;
+    fn emit(&mut self) -> Result<Vec<ArrayRef>>;
+    fn is_empty(&self) -> bool;
+}
+
+// TODO: macro to create for all types, see new_group_values
+// TODO: solve combinatorial issue to not explode compiled binary size
+struct StringInt64Aggregator {
+    descending: bool,
+    random_state: RandomState,
+    priority_map: PriorityMap<String, i64>,
+}
+
+impl StringInt64Aggregator {
+    pub fn new(limit: usize, descending: bool) -> Result<Self> {
+        Ok(Self {
+            descending,
+            random_state: Default::default(),
+            priority_map: PriorityMap::new(limit),
+        })
+    }
+}
+
+impl LimitedAggregator for StringInt64Aggregator {
+    fn intern(&mut self, ids: ArrayRef, vals: &[ArrayRef]) -> Result<()> {
+        let vals = match vals {
+            [] => Err(DataFusionError::Execution("ID column required".to_string()))?,
+            [vals] => vals,
+            _ => Err(DataFusionError::Execution(
+                "One ID column required".to_string(),
+            ))?,
+        };
+        let mut hashes = vec![0; ids.len()];
+        let cols: Vec<ArrayRef> = vec![ids.clone()];
+        create_hashes(cols.as_slice(), &self.random_state, &mut hashes)?;
+
+        let ids = ids.as_any().downcast_ref::<StringArray>().ok_or_else(|| {
+            DataFusionError::Execution("Expected StringArray".to_string())
+        })?;
+        let vals = vals.as_any().downcast_ref::<Int64Array>().ok_or_else(|| {
+            DataFusionError::Execution("Expected Int64Array".to_string())
+        })?;
+        for row_idx in 0..ids.len() {
+            let id = ids.value(row_idx); // TODO: aggregate NULL id
+            let hash = hashes[row_idx];
+            let val = vals.value(row_idx);
+
+            if self.descending {
+                // TODO: val = Reverse(val)
+            }
+            self.priority_map.insert(id.to_string(), hash, val)?;
```

Comment:
It would be nice to avoid the allocation here (`.to_string()`) when not needed, i.e. by passing a `Into<T>` in `insert` and for string pass a `&str` instead, so upstream it can be cloned into the priority map only if needed.

Reply:
Brilliant! I'll do that.

Reply:
As `ID` implements `Clone` I think you can actually also just accept a `&ID` and clone it.

Reply:
I was leaving the door open for custom stuff like `Row` that needs to be converted via `row.owned()` into an `OwnedRow` which didn't fit that pattern.